hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/RingbufferService.java

```diff
@@ -303,4 +328,130 @@ private void clearRingbuffersHavingLesserBackupCountThan(int partitionId, int th
+    }
+
+    private class Merger implements Runnable {
+
+        private static final int TIMEOUT_FACTOR = 500;
+
+        private final Map<Integer, Map<ObjectNamespace, RingbufferContainer>> partitionContainerMap;
+
+        Merger(Map<Integer, Map<ObjectNamespace, RingbufferContainer>> partitionContainerMap) {
+            this.partitionContainerMap = partitionContainerMap;
+        }
+
+        @Override
+        public void run() {
+            // we cannot merge into a 3.9 cluster, since not all members may understand the MergeOperation
+            if (nodeEngine.getClusterService().getClusterVersion().isLessThan(Versions.V3_10)) {
+                return;
+            }
+
+            final ILogger logger = nodeEngine.getLogger(RingbufferService.class);
+            final Semaphore semaphore = new Semaphore(0);
+
+            ExecutionCallback<Object> mergeCallback = new ExecutionCallback<Object>() {
+                @Override
+                public void onResponse(Object response) {
+                    semaphore.release(1);
+                }
+
+                @Override
+                public void onFailure(Throwable t) {
+                    logger.warning("Error while running merge operation: " + t.getMessage());
+                    semaphore.release(1);
+                }
+            };
+
+            int itemCount = 0;
+            int operationCount = 0;
+            List<SplitBrainMergeEntryView<Long, Object>> mergeEntries;
+            for (Entry<Integer, Map<ObjectNamespace, RingbufferContainer>> entry : partitionContainerMap.entrySet()) {
+                int partitionId = entry.getKey();
+                Map<ObjectNamespace, RingbufferContainer> containerMap = entry.getValue();
+
+                for (Entry<ObjectNamespace, RingbufferContainer> containerEntry : containerMap.entrySet()) {
+                    String name = containerEntry.getKey().getObjectName();
+                    RingbufferContainer container = containerEntry.getValue();
+                    Ringbuffer ringbuffer = container.getRingbuffer();
+                    int batchSize = container.getConfig().getMergePolicyConfig().getBatchSize();
+                    SplitBrainMergePolicy mergePolicy = getMergePolicy(container);
+
+                    mergeEntries = new ArrayList<SplitBrainMergeEntryView<Long, Object>>(batchSize);
+                    for (long sequence = ringbuffer.headSequence(); sequence <= ringbuffer.tailSequence(); sequence++) {
+                        Object item = ringbuffer.read(sequence);
+                        SplitBrainMergeEntryView<Long, Object> entryView = createSplitBrainMergeEntryView(sequence, item);
+                        mergeEntries.add(entryView);
+                        itemCount++;
+
+                        if (mergeEntries.size() == batchSize) {
+                            sendBatch(partitionId, name, mergePolicy, mergeEntries, mergeCallback);
```

Comment:
You have to provide the entire `ObjectNamespace` (`containerEntry.getKey()`) to the merge operation as otherwise the `MergeOperation` will [implicitly create a ringbuffer for the ringbuffer service](https://github.com/hazelcast/hazelcast/blob/066d9060cf4c1899d9b4c60db3a20499f5355de8/hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/AbstractRingBufferOperation.java#L76). Since 3.9, there can be ringbuffers for other services as well (event journal for map and cache service)

Reply:
Are you sure? Where should I use that? `MergeOperation extends AbstractRingBufferOperation` which has a `public AbstractRingBufferOperation(String name)` constructor, no `ObjectNameSpace` constructor. Also the code you linked does it's lookup via a `String`, not an `ObjectNameSpace`.

Reply:
That's because event journal entries are never written by the user by invoking ringbuffer operations but by invoking map and cache operations such as `map.put()`. The record store [writes to the ringbuffer behind the scenes](https://github.com/hazelcast/hazelcast/blob/5f9ec24b254dc1cf2442589cae05708542db33f8/hazelcast/src/main/java/com/hazelcast/map/impl/journal/RingbufferMapEventJournalImpl.java#L158-L171).
That's why the `ReplicationOperation` [migrates namespaces as well](https://github.com/hazelcast/hazelcast/blob/fd97a77216e33b5c65a6374ce57a281d9ff5a717/hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReplicationOperation.java#L53-L57).

Reply:
I still have no idea what I should do instead :)

Reply:
The `MergeOperation` has been fixed, I've also added a `MapJournalSplitBrainTest`.